libs/src/sdk/services/Storage/StorageNodeSelector.test.ts

```diff
@@ -0,0 +1,132 @@
+}
+
+const userWallet = '0xc0ffee254729296a45a3885639AC7E10F9d54979'
+
+const discoveryNode = 'https://discovery-provider.audius.co'
+
+class MockAuth implements AuthService {
+  getSharedSecret = async () => new Uint8Array()
+  signTransaction: (data: EIP712TypedData) => Promise<string> = async () => ''
+  sign: (data: string) => Promise<[Uint8Array, number]> = async () => [
+    new Uint8Array(),
+    0
+  ]
+
+  getAddress = async () => {
+    return userWallet
+  }
+}
+
+const auth = new MockAuth()
+
+const mswHandlers = [
+  rest.get(`${discoveryNode}/health_check`, (_req, res, ctx) => {
+    const data: HealthCheckResponseData = {
+      service: 'discovery-node',
+      version: '1.2.3',
+      block_difference: 0,
+      network: {
+        discovery_nodes: [discoveryNode],
+        content_nodes: [storageNodeA, storageNodeB]
+      }
+    }
+
+    return res(ctx.status(200), ctx.json({ data }))
+  }),
+
+  rest.get(`${storageNodeA.endpoint}/status`, (_req, res, ctx) => {
+    return res(ctx.status(200))
+  }),
+
+  rest.get(`${storageNodeB.endpoint}/status`, (_req, res, ctx) => {
+    return res(ctx.status(200))
+  })
+]
+
+const server = setupServer(...mswHandlers)
+
+describe('StorageNodeSelector', () => {
+  beforeAll(() => {
+    server.listen()
+    jest.spyOn(console, 'warn').mockImplementation(() => {})
+    jest.spyOn(console, 'info').mockImplementation(() => {})
+    jest.spyOn(console, 'debug').mockImplementation(() => {})
+    jest.spyOn(console, 'error').mockImplementation(() => {})
+  })
+
+  afterEach(() => {
+    server.resetHandlers()
+  })
+
+  afterAll(() => {
+    server.close()
+  })
+
+  it('selects the correct endpoint given bootstrap nodes and user wallet', async () => {
+    const bootstrapNodes = [storageNodeA, storageNodeB]
+
+    const storageNodeSelector = new StorageNodeSelector({
+      bootstrapNodes,
+      auth
+    })
+
+    expect(await storageNodeSelector.getSelectedNode()).toEqual(
+      storageNodeA.endpoint
+    )
+  })
+
+  it('selects the first healthy node', async () => {
+    server.use(
+      rest.get(`${storageNodeA.endpoint}/status`, (_req, res, ctx) => {
+        return res(ctx.status(400))
+      })
+    )
+    const bootstrapNodes = [storageNodeA, storageNodeB]
+
+    const storageNodeSelector = new StorageNodeSelector({
+      bootstrapNodes,
+      auth
+    })
+
+    expect(await storageNodeSelector.getSelectedNode()).toEqual(
+      storageNodeB.endpoint
+    )
+  })
+
+  it('selects correct storage node when discovery node is selected', async () => {
+    const bootstrapDiscoveryNodes = [discoveryNode]
+    const discoveryNodeSelector = new DiscoveryNodeSelector({
+      bootstrapServices: bootstrapDiscoveryNodes
+    })
+
+    const storageNodeSelector = new StorageNodeSelector({
+      discoveryNodeSelector,
+      auth
+    })
+
+    await discoveryNodeSelector.getSelectedEndpoint()
+
+    await waitForExpect(async () => {
```

Comment:
We might also want to test when DN changes that CN gets the update

Reply:
isn't this testing storage? just using an actual dn selector instead of mocking to be as accurate as possible?

Reply:
I think as long as we have tests for DN selector we can mock out that service and instead just test the Storage parts. Assuming the DN selector has a selected DN, does the storage part work? Assuming the DN selector changes the selected DN, does the storage service get the change? Assuming DN selector doesn't have a healthy DN (returns null) do we handle that appropriately in the storage selector?

If we use a real DN selector we don't get that level of control and we also will have cascading failures (ie Storage tests break if DN selectors break!) and redundant tests

Reply:
would love to chat about this more, because i personally prefer using the real thing where possible, since mocking brings in a host of potential bugs where its misconfigured, but thats more from my perspective with react side which is to use as much real stuff as possible (obvs still have good boundaries around what is actually being tested)

The other thing about using this is if disc-node-selector impl changes itll be easy to forget to update mocks, and then storage-node-selector might break with the transitive change.

just my initial thoughts :)